### ISSUE TITLE
client.query_api().query_data_frame() takes significantly longer than via Data Explorer

### ISSUE THREAD

**Issue by nsbarrow** [influxdata/influxdb-client-python#214]
__Steps to reproduce:__
List the minimal actions needed to reproduce the behavior.

Write a query in Python to get some data into a Pandas dataframe, in this case I have 3 tables adding up to around 11000 rows:
query= '''
from(bucket: "home_assistant")
  |> range(start: 2021-01-23)
  |> filter(fn: (r) => r["friendly_name"] == "Azimuth" or r["friendly_name"] == "Elevation" or r["friendly_name"] == "solaredge (Solar Power)")
  |> filter(fn: (r) => r["_field"] == "value")
  |> aggregateWindow(every: 15m, fn: mean, createEmpty: false)'''

client = InfluxDBClient(url="http://192.168.0.104:8086", token=my_token, org=my_org, debug=False)
df = client.query_api().query_data_frame(org=my_org, query=query)

This query is getting longer everyday, currently around 75 seconds. Conversely, exactly the same query in the Data Explorer web interface is returned in <0.3 s.

__Expected behavior:__
The pandas dataframe is returned in a similar time to the Data Explorer

__Actual behavior:__
The query takes over a minute to return a dataframe.

__Specifications:__
 - Client Version: 1.14.0
 - InfluxDB Version: 2.0.3
 - Platforms: Linux rpi4 4.19.97-v7l+ #1294 SMP Thu Jan 30 13:21:14 GMT 2020 armv7l GNU/Linux
and Linux hass 5.10.11-v8+ #1399 SMP PREEMPT Thu Jan 28 12:14:03 GMT 2021 aarch64

**Comment by bednar** [comment]
Hi @nsbarrow,

thanks for using our client. 

Could you share detail information how your data looks like?  We will try figure out where is a problem. Something like: `print(df.head())` will be pretty enough.

Regards

**Comment by nsbarrow** [comment]
Thanks @bednar , of course, from Python:

`>>> df.info()
<class 'pandas.core.frame.DataFrame'>
Int64Index: 12000 entries, 0 to 11999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype                  
---  ------         --------------  -----                  
 0   result         12000 non-null  object                 
 1   table          12000 non-null  object                 
 2   _start         12000 non-null  datetime64[ns, tzutc()]
 3   _stop          12000 non-null  datetime64[ns, tzutc()]
 4   _time          12000 non-null  datetime64[ns, tzutc()]
 5   _value         12000 non-null  float64                
 6   _field         12000 non-null  object                 
 7   _measurement   12000 non-null  object                 
 8   domain         12000 non-null  object                 
 9   entity_id      12000 non-null  object                 
 10  friendly_name  12000 non-null  object                 
 11  source         12000 non-null  object                 
dtypes: datetime64[ns, tzutc()](3), float64(1), object(8)
memory usage: 843.8+ KB`

`>>> df
        result table  ...            friendly_name source
0      _result     0  ...  solaredge (Solar Power)     HA
1      _result     0  ...  solaredge (Solar Power)     HA
2      _result     0  ...  solaredge (Solar Power)     HA
3      _result     0  ...  solaredge (Solar Power)     HA
4      _result     0  ...  solaredge (Solar Power)     HA
...        ...   ...  ...                      ...    ...
11995  _result     2  ...                Elevation     HA
11996  _result     2  ...                Elevation     HA
11997  _result     2  ...                Elevation     HA
11998  _result     2  ...                Elevation     HA
11999  _result     2  ...                Elevation     HA

[12000 rows x 12 columns]`

The raw annotated CSV from Data Explorer looks like:
`#group,false,false,true,true,false,false,true,true,true,true,true,true
#datatype,string,long,dateTime:RFC3339,dateTime:RFC3339,dateTime:RFC3339,double,string,string,string,string,string,string
#default,_result,,,,,,,,,,,
,result,table,_start,_stop,_time,_value,_field,_measurement,domain,entity_id,friendly_name,source
,,0,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-01-23T16:00:00Z,0.53,value,kW,sensor,solaredge_solar_power,solaredge (Solar Power),HA
,,0,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-01-23T16:15:00Z,0.46,value,kW,sensor,solaredge_solar_power,solaredge (Solar Power),HA
,,0,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-01-23T16:30:00Z,0.19,value,kW,sensor,solaredge_solar_power,solaredge (Solar Power),HA
,,0,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-01-23T16:45:00Z,0,value,kW,sensor,solaredge_solar_power,solaredge (Solar Power),HA
,,0,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-01-24T11:30:00Z,0.01,value,kW,sensor,solaredge_solar_power,solaredge (Solar Power),HA
   ... many skipped lines ...
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T06:45:00Z,4.985714285714286,value,°,sensor,solarelev,Elevation,HA
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T07:00:00Z,7.265,value,°,sensor,solarelev,Elevation,HA
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T07:15:00Z,9.399999999999999,value,°,sensor,solarelev,Elevation,HA
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T07:30:00Z,11.6675,value,°,sensor,solarelev,Elevation,HA
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T07:45:00Z,14.067499999999999,value,°,sensor,solarelev,Elevation,HA
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T08:00:00Z,16.43,value,°,sensor,solarelev,Elevation,HA
,,2,2021-01-23T00:00:00Z,2021-03-22T08:05:44.671483684Z,2021-03-22T08:05:44.671483684Z,17.88,value,°,sensor,solarelev,Elevation,HA`

Thanks for any help you can provide to make the Python client query similar in speed to via Data Explorer.

**Comment by bednar** [comment]
I just tried `query_data_frame()` with following code and [line_protocols.txt](https://github.com/influxdata/influxdb-client-python/files/6182523/line_protocols.txt):

```python
import time
from influxdb_client import InfluxDBClient

client = InfluxDBClient(url="http://localhost:8086", token="my-token", org="my-org")

"""
Query
"""
start_time = time.time()
df = client.query_api().query_data_frame(org="my-org", query='from(bucket: "my-bucket") |> range(start: 0)')
print(df)
end_time = time.time()

"""
Perf
"""
time_lapsed = end_time - start_time
print()
print(f"Elapsed seconds: {time_lapsed}")

"""
Close client
"""
client.close()

```

and everything looks good. 

```
python3 query_perf_python.py 
        result table                    _start                            _stop  ...  domain              entity_id            friendly_name source
0      _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
1      _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
2      _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
3      _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
4      _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
...        ...   ...                       ...                              ...  ...     ...                    ...                      ...    ...
49995  _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
49996  _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
49997  _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
49998  _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA
49999  _result     0 1970-01-01 00:00:00+00:00 2021-03-22 13:58:35.077657+00:00  ...  sensor  solaredge_solar_power  solaredge (Solar Power)     HA

[50000 rows x 12 columns]

Elapsed seconds: 11.695981740951538
```

The most critical part of parsing a response is parsing timestamps. It is recommended to use `ciso8601` for speed up of parsing query response. Do you have installed the `ciso8601` library?

https://github.com/influxdata/influxdb-client-python#installation

#### Client Runtime:
- Linux raspberrypi 5.4.51-v7+ #1333 SMP Mon Aug 10 16:45:19 BST 2020 armv7l GNU/Linux
- Python 3.7.3

**Comment by nsbarrow** [comment]
Thanks for generating some data and testing the client. I have just successfully installed ciso8601-2.1.3 and influxdb-client-1.15.0 (Python 3.7.3), however the runtime for my 12000 rows was only reduced to 53 seconds.

You also gave a long runtime of 11 seconds, compared to the 0.3 s in Data Explorer... is that just the expected behaviour for working with Pandas and maybe not a bug? If so, I'm happy to close the issue.

I also tried these three query types:

`tables = client.query_api().query(query)`

`df = client.query_api().query_data_frame(query)`

```csv_result = client.query_api().query_csv(query,
dialect=Dialect(header=False, delimiter=",", comment_prefix="#", annotations=[],
date_time_format="RFC3339"))```

Tables and df took ~50 s each, and csv_result took <1 s.
But then getting the csv_result into a list took 50 s as well:
`for row in csv_result: mytable.append(row)`

So I'm guessing my bottleneck is CPU constrained to perform the row-by-row operations. I couldn't see an easy way to get the csv_result into a Dataframe. Thanks again for your investigations @bednar !

**Comment by bednar** [comment]
> Thanks for generating some data and testing the client. I have just successfully installed ciso8601-2.1.3 and influxdb-client-1.15.0 (Python 3.7.3), however the runtime for my 12000 rows was only reduced to 53 seconds.
> 
> You also gave a long runtime of 11 seconds, compared to the 0.3 s in Data Explorer... is that just the expected behaviour for working with Pandas and maybe not a bug? If so, I'm happy to close the issue.

Yes, there will be always overhead, because we need to parse incoming data into `datetime`, `int`, `string` and so on.

You could also use a following tweeks to little bit speed up processing:
1. enable `gzip`: `InfluxDBClient(url="http://localhost:8086", token="my-token", org="my-org", enable_gzip=True)`
2. drop unnecessary columns: `query='from(bucket: "my-bucket") |> range(start: 0) |> drop(columns: ["_start", "_stop", "table"])'`

> I also tried these three query types:
> 
> `tables = client.query_api().query(query)`
> 
> `df = client.query_api().query_data_frame(query)`
> 
> ```
> csv_result = client.query_api().query_csv(query,
> dialect=Dialect(header=False, delimiter=",", comment_prefix="#", annotations=[],
> date_time_format="RFC3339"))
> ```
> 
> Tables and df took ~50 s each, and csv_result took <1 s.
> But then getting the csv_result into a list took 50 s as well:
> `for row in csv_result: mytable.append(row)`
> 
> So I'm guessing my bottleneck is CPU constrained to perform the row-by-row operations. I couldn't see an easy way to get the csv_result into a Dataframe. 

You could use something like:

```python
dialect = Dialect(header=False, delimiter=",", comment_prefix="#", annotations=[],
                  date_time_format="RFC3339")

query = 'from(bucket: "my-bucket") ' \
        '|> range(start: 0) ' \
        '|> drop(columns: ["_start", "_stop", "table"])'
csv = client.query_api().query_csv(org="my-org",
                                   query=query,
                                   dialect=dialect)

df = pd.DataFrame(csv)
print(df)
```

> Thanks again for your investigations @bednar !

**Comment by nsbarrow** [comment]
Excellent suggestions @bednar , thank you once again.

The drop columns was most effective:
`|> drop(columns: ["_field", "_measurement", "result", "table", "_start", "_stop", "_field", "domain", "source"])`
halved the query time, down to 25 seconds. Result and table couldn't be dropped though:
```
 0   result         12233 non-null  object             
 1   table          12233 non-null  object             
 2   _time          12233 non-null  datetime64[ns, UTC]
 3   _value         12233 non-null  float64            
 4   entity_id      12233 non-null  object             
 5   friendly_name  12233 non-null  object  
```

When I added `enable_gzip=True` to the client line, I just got an empty dataframe returned after 4 seconds, so that didn't work for me. However, I'm happy to carry on with the dropped columns, that was a huge boost.
